lib/tests/test_rewrite.rs

```diff
@@ -1889,3 +1893,146 @@ fn test_commit_with_selection() {
+    let store = repo.store();
+    // We want to create the following tree with divergent changes A, B, and C:
+    //     E
+    //     |
+    // A2  C2
+    // |   |
+    // C1  B2
+    // |   |
+    // B1  D
+    // |  /
+    // | /
+    // A1
+
+    let tree_a1 = create_tree(repo, &[(repo_path("file1"), "a\n")]);
+    // Commit b deletes file1 and adds file2
+    let tree_b1 = create_tree(repo, &[(repo_path("file2"), "b\n")]);
+    // Commit c appends to file2
+    let tree_c1 = create_tree(repo, &[(repo_path("file2"), "b\nc\n")]);
+    // Commit a2 re-adds file1, the same as commit a1. Since it already has a1
+    // as an ancestor, it shouldn't be abandoned even though it is divergent.
+    let tree_a2 = create_tree(
+        repo,
+        &[(repo_path("file1"), "a\n"), (repo_path("file2"), "b\nc\n")],
+    );
+    let tree_d = create_tree(
+        repo,
+        &[(repo_path("file1"), "a\n"), (repo_path("file3"), "d\n")],
+    );
+    let tree_b2 = create_tree(
+        repo,
+        &[(repo_path("file2"), "b\n"), (repo_path("file3"), "d\n")],
+    );
+    let tree_c2 = create_tree(
+        repo,
+        &[(repo_path("file2"), "b\nc\n"), (repo_path("file3"), "d\n")],
+    );
+    let tree_e = create_tree(
+        repo,
+        &[
+            (repo_path("file2"), "b\nc\n"),
+            (repo_path("file3"), "d\ne\n"),
+        ],
+    );
+
+    let mut make_commit = |tree_id, change_id, parents| {
+        tx.repo_mut()
+            .new_commit(parents, tree_id)
+            .set_change_id(change_id)
+            .write()
+            .unwrap()
+    };
+
+    let commit_a1 = make_commit(
+        tree_a1.id(),
+        ChangeId::from_hex("aaaa"),
+        vec![store.root_commit_id().clone()],
+    );
+    let commit_b1 = make_commit(
+        tree_b1.id(),
+        ChangeId::from_hex("bbbb"),
+        vec![commit_a1.id().clone()],
+    );
+    let commit_c1 = make_commit(
+        tree_c1.id(),
+        ChangeId::from_hex("cccc"),
+        vec![commit_b1.id().clone()],
+    );
+    let commit_a2 = make_commit(
+        tree_a2.id(),
+        ChangeId::from_hex("aaaa"),
+        vec![commit_c1.id().clone()],
+    );
+    let commit_d = make_commit(
+        tree_d.id(),
+        ChangeId::from_hex("dddd"),
+        vec![commit_a1.id().clone()],
+    );
+    let commit_b2 = make_commit(
+        tree_b2.id(),
+        ChangeId::from_hex("bbbb"),
+        vec![commit_d.id().clone()],
+    );
+    let commit_c2 = make_commit(
+        tree_c2.id(),
+        ChangeId::from_hex("cccc"),
+        vec![commit_b2.id().clone()],
+    );
+    let commit_e = make_commit(
+        tree_e.id(),
+        ChangeId::from_hex("eeee"),
+        vec![commit_c2.id().clone()],
+    );
+
+    #[track_caller]
+    fn assert_same_ids(commits: &[Commit], ids: &[&CommitId]) {
+        assert_eq!(
+            commits.iter().map(Commit::id).sorted().collect_vec(),
+            ids.iter().sorted().copied().collect_vec(),
+        );
+    }
```

Comment:
nit: maybe we can just `assert_eq!(&[Commit], &[Commit])`? I think the order of the returned list is deterministic.